### PR TITLE
Fix tool.SBOM.cyclonedx with empty components

### PR DESCRIPTION
**Pull request by NokiDev**
Changelog: BugFix: Add correct info in metadata + prevent crash when no component is associated to root_node.
Docs: Omit

Fix issue #17924

- [x] Refer to the issue that supports this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop2/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one.

### CODE REVIEW

File: conan/tools/sbom/cyclonedx.py [hunk 1]
```diff
@@ -61,9 +71,12 @@ def _calculate_licenses(component):
                     "id": l
                 }} for l in component.conanfile.license]
 
+    if not components:
+        components = {}
+
     sbom_cyclonedx_1_4 = {
         **({"components": [{
-            "author": "Conan",
+            **({"author": c.conanfile.author} if c.conanfile.author else {}),
```
Comment (NokiDev) [inline conan/tools/sbom/cyclonedx.py]: Not sure why it was "Conan" here.

**Comment by ErniGH** [comment]
Hi @NokiDev, I’ve been looking into the issue, and there is indeed a problem. The [line](https://github.com/conan-io/conan/pull/17925/files#diff-3bfd3bcbe61f8f8ccff41094f7d1e1899e2ba20bbeb3ea3f7b6658045ef7eff4R96) that fixes it is:
 ` 
 "bom-ref": special_id if has_special_root_node else f"pkg:conan/{conanfile.name}@{conanfile.ref.version}?rref={conanfile.ref.revision}",
`
 so I think it would be best to minimize the pull request, focusing on fixing the issue so we can merge it and resolve the bug.

I see too many changes unrelated to the main issue, and I believe it would be better to discuss those in a separate PR.

Regarding the author field, it is indeed incorrect to use "conan" since the standard refers to "The person(s) or organization(s) that authored the component." I think it would be best to remove it altogether, as it is not a mandatory field in the standard, and it is not filled in the Conan Center Index recipes, which could cause issues.

**Review by ErniGH — APPROVED** [review]
LGTM Thanks for your PR 😄

**Comment by NokiDev** [comment]
@ErniGH, I've followed your advice and limited PR change to the fix.
I'm open to discuss all other changes in another issue / PR if there is one yet

**Comment by ErniGH** [comment]
@NokiDev I feel a bit uneasy about not having had a test that could catch this error. I'm trying to add a small test to cover this case, but I'm unable to replicate the issue you had when the component list was empty. Could you help me? What was your use case for triggering an empty component list?

**Comment by NokiDev** [comment]
Sure, in my case I was using a conanfile with tool_requires and calling the function with use_build=False (the default I believe) 

But without any requires that should also work no ? 
Otherwise having a test that check the outputed spdx.json file content shall be better. 
And then you can check that in the metadata section you have intended info ? 
IMO better to check intended output than all error ones. 

If you still struggle, I can add a test tomorrow - not sure how testing a hook with conan TestClient tough

**Comment by ErniGH** [comment]
I have added a test that has a "post_package" hook with `add_build` set to `false` and `add_test` set to `false` as well. I create a `conanfile.py` without any requirements and build it using `conan create`. This test does not generate an empty component list in the SBOM. How did you do it? 🤔

File: test/functional/sbom/test_cyclonedx.py [hunk 2]
```diff
@@ -230,6 +262,24 @@ def test_sbom_generation_install_path_txt(hook_setup_post_generate):
...
+def test_sbom_with_special_root_node(hook_setup_post_generate):
+    tc = hook_setup_post_generate
+    package_name = "foo"
+
+    conanfile =  textwrap.dedent("""
+            from conan import ConanFile
+            class FooPackage(ConanFile):
+                name = "foo"
+                version = "1.0"
+                package_type = "build-scripts"
```
Comment (NokiDev) [inline test/functional/sbom/test_cyclonedx.py:250]: I achieved to reproduce the issue when there is no components by setting package-type = "build-scripts" 
Not really sure why, it triggers correctly the variable has_special_root_node which in turn would strip 1st component from the graph and triggers the issue.

File: test/functional/sbom/test_cyclonedx.py [hunk 3]
```diff
@@ -66,6 +72,32 @@ def test_sbom_generation_create(hook_setup_post_package_tl):
     bar_layout = tc.created_layout()
     assert os.path.exists(os.path.join(bar_layout.metadata(), "sbom.cdx.json"))
 
+def test_sbom_with_dependencies(hook_setup_post_package_default):
```
Comment (NokiDev) [inline test/functional/sbom/test_cyclonedx.py]: Modified this test to ensure sbom is generated with appropriate component ref in the metadata.

**Comment by NokiDev** [comment]
@ErniGH Achieved to reproduce, I believe I might have tried to use the post_generate hook too when issue occured. My bad.

I updated the tests to reflect the changes I made.

BTW if you have a less weird way to trigger the has_special_root_node behaviour, feel free to update the test. I found that having package_type set as build scripts works. IMHO stripping the 1st component shall be the default behaviour since information will be present in the metadata section

Edit: This seems really weird because has_special_root_node is True, thus the special id is used in bom-ref and it shouldn't execute the else clause where c.name etc... are used but it seems to evaluate the else clause as well.
Is this a python interpreter normal behaviour ?

**Comment by ErniGH** [comment]
@NokiDev The problem here is that the subgraph generated of a `build-script` without requirements generates a single node, the context of which is built and is not used by the SBOM. The test is able to capture the error, and the fix resolves the bug.
Good job!

**Review by ErniGH — APPROVED** [review]
LGTM!

**Review by memsharded — APPROVED** [review]
